apps/emqx/src/emqx_limiter/src/emqx_limiter_shared/emqx_limiter_shared.erl

```diff
@@ -61,101 +98,264 @@
     {emqx_limiter:name(), emqx_limiter:options()}
 ]) -> ok | {error, term()}.
 create_group(Group, LimiterConfigs) when length(LimiterConfigs) > 0 ->
-    {Names, _} = lists:unzip(LimiterConfigs),
-    ChildSpec = child_spec(Group, Names),
-    start_child(emqx_limiter_shared_sup, ChildSpec).
+    Size = length(LimiterConfigs),
+    MiniTokensCRef = counters:new(Size, [write_concurrency]),
+    LastTimeARef = atomics:new(Size * 2, []),
+    NowUs = now_us_monotonic(),
+    Buckets = make_buckets(LimiterConfigs, MiniTokensCRef, LastTimeARef, NowUs),
+    ok = emqx_limiter_bucket_registry:insert_buckets(Group, Buckets).
 
 -spec delete_group(emqx_limiter:group()) -> ok | {error, term()}.
 delete_group(Group) ->
-    case supervisor:terminate_child(emqx_limiter_shared_sup, Group) of
-        ok ->
-            supervisor:delete_child(emqx_limiter_shared_sup, Group);
-        {error, not_found} ->
-            ok
-    end.
+    ok = emqx_limiter_bucket_registry:delete_buckets(Group).
 
 -spec update_group(emqx_limiter:group(), [{emqx_limiter:name(), emqx_limiter:options()}]) ->
-    ok | no_return().
-update_group(Group, _LimiterConfigs) ->
-    ok = emqx_limiter_allocator:update(Group).
-
--spec child_spec(emqx_limiter:group(), [emqx_limiter:name()]) ->
-    supervisor:child_spec().
-child_spec(Group, Names) ->
-    #{
-        id => Group,
-        start => {emqx_limiter_allocator, start_link, [Group, Names]},
-        restart => permanent,
-        shutdown => 5000,
-        type => worker,
-        modules => [?MODULE, emqx_limiter_allocator]
-    }.
+    ok.
+update_group(_Group, _LimiterConfigs) ->
+    ok.
 
 -spec connect(emqx_limiter:id()) -> emqx_limiter_client:t().
 connect({Group, Name}) ->
```

Comment:
maybe there is no need for a 'container' module if connect accepts a list of IDs?
and there is no need to expose the put_back api.

Reply:
The containers currently combine different types of limiters.

Reply:
ok,.
I'm thinking about the multi-tenancy case, each client may have two shared limiters 
1. tenant level
2. node level
should we make a wrapping container for the two shared limiters ?

Reply:
There are two dimensions of containerization.
One is combining scopes: node limiter, zone limiter, tenant limiter, etc. 
The other is combining different units: bytes limiter, message count limiter.
If we consume bytes but don't consume messages, we should put messages back.

We probably may put all these inside a single limiter client, but
* we would still have to somehow combine such limiters with exclusive ones (or deprecate them?)
* we actually cannot consume atomically several limiters, so the same algo of putting back will be just hidden
* the put_back thing is already not exposed to the client (channel); it operates with a container.

Reply:
> the put_back thing is already not exposed to the client (channel); a channel operates with a container.

Yes this I understand, what I meant is the export from this module to container.
Now that I also understand that the container is necessary, so exporting put_back is fine.
Thank your for the explain